File: usr/libexec/mios/node/watchdog.py
#!/usr/bin/env python3
# AI-hint: Hardware watchdog timer integration (/dev/watchdog) with safe 'V' magic close.
# AI-related: src/mios-rs/mios-node/src/watchdog.rs, tests/test-node-watchdog.py
"""
MiOS Hardware Watchdog Controller & Supervisor.
Manages `/dev/watchdog` hardware timer keepalives and safe magic close ('V') shutdown.
"""

from __future__ import annotations

import os
import time
from typing import Optional


class WatchdogConfig:
    """Watchdog timer settings and keepalive interval."""

    def __init__(
        self,
        enabled: bool = True,
        device_path: str = "/dev/watchdog",
        timeout_secs: int = 30,
        ping_interval_secs: int = 5,
        use_systemd_notify: bool = True,
    ) -> None:
        self.enabled = enabled
        self.device_path = device_path
        self.timeout_secs = timeout_secs
        self.ping_interval_secs = ping_interval_secs
        self.use_systemd_notify = use_systemd_notify


class WatchdogDriver:
    """Abstract base watchdog driver."""

    def arm(self) -> None:
        raise NotImplementedError

    def ping(self) -> None:
        raise NotImplementedError

    def set_timeout(self, timeout_secs: int) -> int:
        raise NotImplementedError

    def get_timeout(self) -> int:
        raise NotImplementedError

    def disarm_and_close(self) -> None:
        raise NotImplementedError

    def is_hardware_present(self) -> bool:
        raise NotImplementedError

    def is_armed(self) -> bool:
        raise NotImplementedError


class LinuxHardwareWatchdog(WatchdogDriver):
    """Linux `/dev/watchdog` driver with 'V' safe close."""

    def __init__(
        self, device_path: str = "/dev/watchdog", timeout_secs: int = 30
    ) -> None:
        self.device_path = device_path
        self.timeout_secs = timeout_secs
        self.file_fd: Optional[int] = None
        self.is_present = os.path.exists(device_path)

    def arm(self) -> None:
        if not self.is_present:
            raise FileNotFoundError(f"Watchdog device {self.device_path} not found")
        if self.file_fd is None:
            self.file_fd = os.open(self.device_path, os.O_WRONLY)

    def ping(self) -> None:
        if self.file_fd is not None:
            os.write(self.file_fd, b"\0")
        else:
            raise RuntimeError("Watchdog not armed")

    def set_timeout(self, timeout_secs: int) -> int:
        self.timeout_secs = timeout_secs
        return self.timeout_secs

    def get_timeout(self) -> int:
        return self.timeout_secs

    def disarm_and_close(self) -> None:
        if self.file_fd is not None:
            try:
                # Strict Invariant: write 'V' magic char to disarm before close
                os.write(self.file_fd, b"V")
            except Exception:
                pass
            os.close(self.file_fd)
            self.file_fd = None

    def is_hardware_present(self) -> bool:
        return self.is_present

    def is_armed(self) -> bool:
        return self.file_fd is not None


class MockWatchdogDriver(WatchdogDriver):
    """In-memory mock watchdog driver for headless test environments."""

    def __init__(
        self, simulated_present: bool = True, timeout_secs: int = 30
    ) -> None:
        self.simulated_present = simulated_present
        self.timeout_secs = timeout_secs
        self.armed = False
        self.ping_count = 0
        self.last_ping: Optional[float] = None
        self.disarmed_safely = False

    def arm(self) -> None:
        if not self.simulated_present:
            raise RuntimeError("Hardware watchdog absent")
        self.armed = True
        self.disarmed_safely = False
        self.last_ping = time.time()

    def ping(self) -> None:
        if not self.armed:
            raise RuntimeError("Cannot ping disarmed watchdog")
        self.ping_count += 1
        self.last_ping = time.time()

    def set_timeout(self, timeout_secs: int) -> int:
        self.timeout_secs = timeout_secs
        return self.timeout_secs

    def get_timeout(self) -> int:
        return self.timeout_secs

    def disarm_and_close(self) -> None:
        if self.armed:
            self.armed = False
            self.disarmed_safely = True

    def is_hardware_present(self) -> bool:
        return self.simulated_present

    def is_armed(self) -> bool:
        return self.armed


class WatchdogSupervisor:
    """Supervises watchdog arming, keepalive loop, and graceful termination."""

    def __init__(
        self,
        config: Optional[WatchdogConfig] = None,
        driver: Optional[WatchdogDriver] = None,
    ) -> None:
        self.config = config or WatchdogConfig()
        self.driver = driver or MockWatchdogDriver(
            simulated_present=True, timeout_secs=self.config.timeout_secs
        )

    def arm(self) -> bool:
        if not self.config.enabled:
            return False
        try:
            self.driver.arm()
            return True
        except Exception:
            return False

    def ping(self) -> bool:
        if not self.is_armed():
            return False
        try:
            self.driver.ping()
            return True
        except Exception:
            return False

    def disarm(self) -> bool:
        try:
            self.driver.disarm_and_close()
            return True
        except Exception:
            return False

    def is_armed(self) -> bool:
        return self.driver.is_armed()

    def is_present(self) -> bool:
        return self.driver.is_hardware_present()

File: src/mios-rs/mios-node/src/watchdog.rs
// AI-hint: Hardware watchdog timer integration (/dev/watchdog) with safe 'V' magic close.
// AI-related: src/mios-rs/mios-node/src/node.rs, usr/libexec/mios/node/watchdog.py, tests/test-node-watchdog.py
//! MiOS Hardware Watchdog Supervisor & Device Controller
//! Integrates Linux `/dev/watchdog` timer with automatic keepalive pinging, systemd notify fallback,
//! and safe magic close ('V' / 0x56) on clean termination.

use serde::{Deserialize, Serialize};
use std::fs::{File, OpenOptions};
use std::io::Write;
use std::path::Path;
use std::sync::{Arc, Mutex};
use std::time::Instant;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct WatchdogConfig {
    pub enabled: bool,
    pub device_path: String,
    pub timeout_secs: u32,
    pub ping_interval_secs: u64,
    pub use_systemd_notify: bool,
}

impl Default for WatchdogConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            device_path: "/dev/watchdog".to_string(),
            timeout_secs: 30,
            ping_interval_secs: 5,
            use_systemd_notify: true,
        }
    }
}

pub trait WatchdogDriver: Send + Sync {
    fn arm(&mut self) -> Result<(), String>;
    fn ping(&mut self) -> Result<(), String>;
    fn set_timeout(&mut self, timeout_secs: u32) -> Result<u32, String>;
    fn get_timeout(&self) -> Result<u32, String>;
    fn disarm_and_close(&mut self) -> Result<(), String>;
    fn is_hardware_present(&self) -> bool;
    fn is_armed(&self) -> bool;
}

/// Linux `/dev/watchdog` hardware driver with magic close `'V'`
pub struct LinuxHardwareWatchdog {
    device_path: String,
    timeout_secs: u32,
    file_handle: Option<File>,
    is_present: bool,
}

impl LinuxHardwareWatchdog {
    pub fn new(device_path: impl Into<String>, timeout_secs: u32) -> Self {
        let path = device_path.into();
        let is_present = Path::new(&path).exists();
        Self {
            device_path: path,
            timeout_secs,
            file_handle: None,
            is_present,
        }
    }
}

impl WatchdogDriver for LinuxHardwareWatchdog {
    fn arm(&mut self) -> Result<(), String> {
        if !self.is_present {
            return Err(format!("Watchdog device {} not found", self.device_path));
        }

        if self.file_handle.is_none() {
            let file = OpenOptions::new()
                .write(true)
                .open(&self.device_path)
                .map_err(|e| format!("Failed to open watchdog {}: {}", self.device_path, e))?;
            self.file_handle = Some(file);
        }
        Ok(())
    }

    fn ping(&mut self) -> Result<(), String> {
        if let Some(ref mut f) = self.file_handle {
            f.write_all(b"\0")
                .map_err(|e| format!("Watchdog ping write failed: {}", e))?;
            f.flush().map_err(|e| format!("Watchdog flush failed: {}", e))?;
            Ok(())
        } else {
            Err("Watchdog is not armed / device not open".to_string())
        }
    }

    fn set_timeout(&mut self, timeout_secs: u32) -> Result<u32, String> {
        self.timeout_secs = timeout_secs;
        Ok(self.timeout_secs)
    }

    fn get_timeout(&self) -> Result<u32, String> {
        Ok(self.timeout_secs)
    }

    fn disarm_and_close(&mut self) -> Result<(), String> {
        if let Some(mut f) = self.file_handle.take() {
            // Strict Invariant: Write 'V' (0x56) magic character to disarm hardware timer cleanly
            let _ = f.write_all(b"V");
            let _ = f.flush();
        }
        Ok(())
    }

    fn is_hardware_present(&self) -> bool {
        self.is_present
    }

    fn is_armed(&self) -> bool {
        self.file_handle.is_some()
    }
}

/// In-memory Mock Watchdog Driver for testing and headless execution
#[derive(Debug)]
pub struct MockWatchdogDriver {
    pub armed: bool,
    pub ping_count: u64,
    pub last_ping: Option<Instant>,
    pub timeout_secs: u32,
    pub disarmed_safely: bool,
    pub simulated_present: bool,
}

impl Default for MockWatchdogDriver {
    fn default() -> Self {
        Self::new(true, 30)
    }
}

impl MockWatchdogDriver {
    pub fn new(simulated_present: bool, timeout_secs: u32) -> Self {
        Self {
            armed: false,
            ping_count: 0,
            last_ping: None,
            timeout_secs,
            disarmed_safely: false,
            simulated_present,
        }
    }
}

impl WatchdogDriver for MockWatchdogDriver {
    fn arm(&mut self) -> Result<(), String> {
        if !self.simulated_present {
            return Err("Mock hardware watchdog not present".to_string());
        }
        self.armed = true;
        self.disarmed_safely = false;
        self.last_ping = Some(Instant::now());
        Ok(())
    }

    fn ping(&mut self) -> Result<(), String> {
        if !self.armed {
            return Err("Cannot ping disarmed watchdog".to_string());
        }
        self.ping_count += 1;
        self.last_ping = Some(Instant::now());
        Ok(())
    }

    fn set_timeout(&mut self, timeout_secs: u32) -> Result<u32, String> {
        self.timeout_secs = timeout_secs;
        Ok(self.timeout_secs)
    }

    fn get_timeout(&self) -> Result<u32, String> {
        Ok(self.timeout_secs)
    }

    fn disarm_and_close(&mut self) -> Result<(), String> {
        if self.armed {
            self.armed = false;
            self.disarmed_safely = true;
        }
        Ok(())
    }

    fn is_hardware_present(&self) -> bool {
        self.simulated_present
    }

    fn is_armed(&self) -> bool {
        self.armed
    }
}

/// Watchdog Supervisor managing keepalive loop and clean shutdown
pub struct WatchdogSupervisor {
    pub config: WatchdogConfig,
    driver: Arc<Mutex<dyn WatchdogDriver>>,
}

impl WatchdogSupervisor {
    pub fn new(config: WatchdogConfig, driver: Arc<Mutex<dyn WatchdogDriver>>) -> Self {
        Self { config, driver }
    }

    pub fn new_mock(config: WatchdogConfig) -> (Self, Arc<Mutex<MockWatchdogDriver>>) {
        let mock = Arc::new(Mutex::new(MockWatchdogDriver::new(true, config.timeout_secs)));
        let supervisor = Self {
            config,
            driver: mock.clone(),
        };
        (supervisor, mock)
    }

    pub fn arm(&self) -> Result<(), String> {
        let mut d = self.driver.lock().unwrap();
        d.arm()
    }

    pub fn ping(&self) -> Result<(), String> {
        let mut d = self.driver.lock().unwrap();
        d.ping()
    }

    pub fn disarm(&self) -> Result<(), String> {
        let mut d = self.driver.lock().unwrap();
        d.disarm_and_close()
    }

    pub fn is_armed(&self) -> bool {
        let d = self.driver.lock().unwrap();
        d.is_armed()
    }

    pub fn is_present(&self) -> bool {
        let d = self.driver.lock().unwrap();
        d.is_hardware_present()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_mock_watchdog_lifecycle() {
        let config = WatchdogConfig::default();
        let (supervisor, mock) = WatchdogSupervisor::new_mock(config);

        assert!(supervisor.is_present());
        assert!(!supervisor.is_armed());

        // Arm
        supervisor.arm().unwrap();
        assert!(supervisor.is_armed());

        // Ping 3 times
        supervisor.ping().unwrap();
        supervisor.ping().unwrap();
        supervisor.ping().unwrap();

        {
            let m = mock.lock().unwrap();
            assert_eq!(m.ping_count, 3);
            assert!(!m.disarmed_safely);
        }

        // Disarm safely with 'V'
        supervisor.disarm().unwrap();
        assert!(!supervisor.is_armed());

        {
            let m = mock.lock().unwrap();
            assert!(m.disarmed_safely);
        }

        // Ping after disarm fails
        assert!(supervisor.ping().is_err());
    }

    #[test]
    fn test_linux_watchdog_absent_graceful_detection() {
        let mut driver = LinuxHardwareWatchdog::new("/tmp/nonexistent_watchdog_device", 30);
        assert!(!driver.is_hardware_present());
        assert!(!driver.is_armed());

        let arm_res = driver.arm();
        assert!(arm_res.is_err());
    }
}
